content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md

```diff
@@ -0,0 +1,808 @@
+[documentation](https://llvm.org/docs/JITLink.html#linkgraph) has an excellent description of LinkGraph). LinkGraph is an internal representation of an object file,
+in LLVM JITLink. All object files, no matter the format and target architecture, are represented as LinkGraphs in JITLink.
+
+Object formats have different schemas and names for similar concepts, but they share a common goal - to represent
+machine code that can be relocated in virtual memory. A LinkGraph tries to generically represent similar concepts and nuances across object files.
+
+To draw conceptual analogies between the LinkGraph and an object format, I will use ELF as an example since it is
+the format I am most familiar with. An ELF object contains sections, symbols and relocations.
+
+1. A section is any chunk of bytes that must be moved into memory as a unit.
+2. A symbol is a named chunk of bytes that could represent either data or executable instructions. Symbols occur as children of sections.
+3. A relocation describes how to fix up bytes within a section once the address of the relocation's target symbol is resolved.
+
+A LinkGraph is capable of representing all of the above concepts. It first defines some building blocks.
+
+1. [Addressable](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L105) - Anything that can be assigned an address in the executor process’ virtual address space.
+2. [Block](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L152) - A chunk of bytes that is addressable and occurs as part of a section.
+
+On top of these building blocks it defines the higher level object format concepts.
+
+1. [Symbol](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L408) - Equivalent of a symbol in the ELF format. Represented using an offset from the base (address) of a
+   Block + a size.
+2. [Section](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L673) - Equivalent of a section in the ELF format. Represented using a collection of symbols and blocks.
+3. [Edge](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L61) - Equivalent of a relocation in the ELF format. Represented using an offset from the start of the
+   containing block (indicating the storage location that needs to be fixed up), a pointer to the target whose address needs to be used for the fix-up and a kind to specify the patching formula.
+
+> Another important LLVM JITLink construct, that I did not directly interact with too much, is [JITLinkContext](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ExecutionEngine/JITLink/JITLink.h#L61). But
+> @lhames suggested and I agreed that this would be a good place to talk about it, since it is one of the building blocks of LLVM JITLink.
```

Comment:
> You could make space by reducing the higher level concepts into one sentence, since they all map to ELF very well.

I think we would be assuming that the reader or a new contributor is already familiar with ELF or object formats in general. For that reason I'd prefer being a bit verbose here.